src/Features/CSharp/Portable/MakeStructFieldsWritable/CSharpMakeStructFieldsWritableDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,102 @@
+
+        private sealed class SymbolAnalyzer
+        {
+            private bool _hasTypeInstanceAssigment = false;
+            private INamedTypeSymbol _namedTypeSymbol;
+            private SymbolAnalyzer(INamedTypeSymbol namedTypeSymbol)
+            {
+                _namedTypeSymbol = namedTypeSymbol;
+            }
+
+            public static void CreateAndRegisterActions(CompilationStartAnalysisContext compilationStartContext)
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    // We report diagnostic only if these requirements are met:
+                    // 1. The type is struct
+                    // 2. Struct contains at least one 'readonly' field
+                    // 3. Struct contains assignment to 'this' outside the scope of constructor
+                    var namedTypeSymbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+
+                    // We are only interested in struct declarations
+                    if (namedTypeSymbol.TypeKind != TypeKind.Struct) return;
+                    //We check if struct contains any 'readonly' fields
+                    if (!HasReadonlyField(namedTypeSymbol)) return;
+
+                    var symbolAnalyzer = new SymbolAnalyzer(namedTypeSymbol);
+                    symbolAnalyzer.RegisterActions(symbolStartContext);
+
+                }, SymbolKind.NamedType);
+            }
+
+            private static bool HasReadonlyField(INamedTypeSymbol namedTypeSymbol)
+            {
+                return namedTypeSymbol
+                .GetMembers()
+                .OfType<IFieldSymbol>()
+                .Where(field => field.AssociatedSymbol == null)
+                .Any(field => field.IsReadOnly);
+            }
+
+            private void RegisterActions(SymbolStartAnalysisContext symbolStartContext)
+            {
+                symbolStartContext.RegisterOperationAction(AnalyzeAssignment, OperationKind.SimpleAssignment);
+                symbolStartContext.RegisterSymbolEndAction(SymbolEndAction);
+            }
+
+            private void AnalyzeAssignment(OperationAnalysisContext operationContext)
+            {
+                // We are looking for assignment to 'this' outside the constructor scope
+                var operationAssigmnent = (IAssignmentOperation)operationContext.Operation;
+                _hasTypeInstanceAssigment = operationAssigmnent.Target is IInstanceReferenceOperation instance &&
+                                            instance.ReferenceKind == InstanceReferenceKind.ContainingTypeInstance;
+            }
+
+            private void SymbolEndAction(SymbolAnalysisContext symbolEndContext)
+            {
+                if (_hasTypeInstanceAssigment)
+                {
+                    var diagnostic = Diagnostic.Create(
+                                    s_diagnosticDescriptor,
+                                    _namedTypeSymbol.Locations[0]);
```

Comment:
there is a disconnect ebtween this diagnostic and your code fixer.  If you intend to get back to a specific syntax node in the fixer, then you need to ensure that you pass a proper location here that you know will be safe to roundtrip.

Reply:
@CyrusNajmabadi 
Thank you for your feedback.

Could you be please more precise and provide some specific suggestion. It's not fully clear to me what should be checked before reporting the diagnostic.

If you are talking about not checking the type `INamedTypeSymbol`, whenever it is struct or class.
The check is done before creating the instance of `SymbolAnalyzer` class.
[https://github.com/dotnet/roslyn/blob/426cfe38960564842e151a3f3dc7de2f80d44baf/src/Features/CSharp/Portable/MakeStructFieldsWritable/CSharpMakeStructFieldsWritableDiagnosticAnalyzer.cs#L56-L59](https://github.com/dotnet/roslyn/blob/426cfe38960564842e151a3f3dc7de2f80d44baf/src/Features/CSharp/Portable/MakeStructFieldsWritable/CSharpMakeStructFieldsWritableDiagnosticAnalyzer.cs#L56-L59)

Is that it, what were you commenting?